crates/oxc_linter/src/rules/typescript/prefer_for_of.rs

```diff
@@ -0,0 +1,296 @@
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::{Atom, GetSpan, Span};
+use oxc_syntax::operator::{AssignmentOperator, BinaryOperator, UnaryOperator, UpdateOperator};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+trait SpanExt {
+    fn contains(&self, other: Self) -> bool;
+}
+
+impl SpanExt for Span {
+    fn contains(&self, other: Self) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+}
+
+trait ExpressionExt {
+    fn is_increment_of(&self, var_name: &Atom) -> bool;
+}
+
+impl<'a> ExpressionExt for Expression<'a> {
+    fn is_increment_of(&self, var_name: &Atom) -> bool {
+        match self {
+            Expression::UpdateExpression(expr) => match (&expr.argument, &expr.operator) {
+                (
+                    SimpleAssignmentTarget::AssignmentTargetIdentifier(id),
+                    UpdateOperator::Increment,
+                ) => id.name == var_name,
+                _ => false,
+            },
+            Expression::AssignmentExpression(expr) => {
+                if !matches!(&expr.left,
+                    AssignmentTarget::SimpleAssignmentTarget(
+                        SimpleAssignmentTarget::AssignmentTargetIdentifier(id)
+                    )
+                    if id.name == var_name
+                ) {
+                    return false;
+                }
+
+                match expr.operator {
+                    AssignmentOperator::Addition => {
+                        matches!(&expr.right, Expression::NumericLiteral(lit)
+                            if (lit.value - 1f64).abs() < f64::EPSILON)
+                    }
+                    AssignmentOperator::Assign => {
+                        let Expression::BinaryExpression(bin_expr) = &expr.right else {
+                            return false;
+                        };
+
+                        if bin_expr.operator != BinaryOperator::Addition {
+                            return false;
+                        }
+
+                        match (&bin_expr.left, &bin_expr.right) {
+                            (Expression::Identifier(id), Expression::NumericLiteral(lit))
+                            | (Expression::NumericLiteral(lit), Expression::Identifier(id)) => {
+                                id.name == var_name && (lit.value - 1f64).abs() < f64::EPSILON
+                            }
+                            _ => false,
+                        }
+                    }
+                    _ => false,
+                }
+            }
+            _ => false,
+        }
+    }
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(prefer-for-of): Expected a `for-of` loop instead of a `for` loop with this simple iteration.")]
+#[diagnostic(severity(warning), help("Consider using a for-of loop for this simple iteration."))]
+struct PreferForOfDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferForOf;
+
+declare_oxc_lint!(
```

Comment:
Can we put these main codes at the top of files?  It would provide a more clear overview of the rule's structure.

Reply:
Done.